managers/asset-operations-manager.js

```diff
@@ -344,71 +345,75 @@ export default class AssetOperationsManager {
             dataset = await kcTools.formatDataset(content);
         }
 
+        let publicTriplesGrouped = [];
+        let tokensCount = 0;
+        // Assign IDs to blank nodes
+
+        dataset.public = kcTools.generateMissingIdsForBlankNodes(dataset.public);
+
         if (dataset.private?.length) {
             dataset.private = kcTools.generateMissingIdsForBlankNodes(dataset.private);
 
+            // Group private triples by subject and flatten
             const privateTriplesGrouped = kcTools.groupNquadsBySubject(dataset.private, true);
             dataset.private = privateTriplesGrouped.flat();
-            const privateRoot = kcTools.calculateMerkleRoot(dataset.private);
 
+            // Compute private root and add to public
+            const privateRoot = kcTools.calculateMerkleRoot(dataset.private);
             dataset.public.push(
                 `<${kaTools.generateNamedNode()}> <${PRIVATE_ASSERTION_PREDICATE}> "${privateRoot}" .`,
             );
 
-            if (dataset.public.length) {
-                dataset.public = kcTools.generateMissingIdsForBlankNodes(dataset.public);
+            // Group public triples by subject
+            publicTriplesGrouped = kcTools.groupNquadsBySubject(dataset.public, true);
+            tokensCount += publicTriplesGrouped.length;
+
+            // Create a map of public subject -> index for quick lookup
+            const publicSubjectMap = new Map();
```

Comment:
```suggestion
            const publicSubjectMap = new Map(
    publicTriplesGrouped.map(([triple], index) => [triple.split(' ')[0], index])
);

```

you can do tihs in one liner very cleanly imo

Reply:
and remove stuff below